### PR TITLE
Mixture-of-Experts partial diffusion implementation for base SD 1.x / 2.x pipelines

### PR DESCRIPTION
**Pull request by bghira**
# What does this PR do?

This pull request ports our denoising_start code to the text2img pipeline, and the denoising_start and denoising_end code from the img2img pipeline.

This brings legacy SD model capabilities in line with SDXL.

Enhances #4003 

# Example

Using on Stable Diffusion 2.1 fine-tuned with zero terminal SNR:

image
image
image
image

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [ ] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by bghira** [comment]
@patrickvonplaten making sure this one doesn't get lost

**Comment by patrickvonplaten** [comment]
Very cool addition! Could we maybe add some tests? Think they can be very similar to the ones we added here: https://github.com/huggingface/diffusers/blob/d0b8de1262ba785474fc9df53c29ba44ec02c715/tests/pipelines/stable_diffusion_xl/test_stable_diffusion_xl.py#L264 (feel free to copy-paste)

**Comment by patrickvonplaten** [comment]
Let us know if you need any help with the tests or currently failing tests @bghira :-)

**Comment by bghira** [comment]
@patrickvonplaten sorry, i've been really busy testing SDXL training and haven't had time to follow-up here. i would be glad for the assist!

**Comment by bghira** [comment]
@patrickvonplaten

**Comment by patrickvonplaten** [comment]
@bghira for which SD 1.x / SD 2.x models does mixture of experts work well?

**Comment by bghira** [comment]
the result above is actually from passing the partially diffused output from SD 2.1, through SDXL.

**Comment by sayakpaul** [comment]
@yiyixuxu WDYT?

**Comment by patrickvonplaten** [comment]
I don't think we need to prio this PR at the moment

**Comment by bghira** [comment]
come on

**Comment by patrickvonplaten** [comment]
@bghira sorry to be so late here. Could you maybe post a code snippet showing how this PR enables high quality images with Mixture Of Experts for SD 1.x and 2.x? Given that we don't have 1.x or 2.x denoiser checkpoint I'm a bit unsure whether this is needed yet to be honest.

**Comment by bghira** [comment]
my trainer can make them, and my 2.x checkpoints can make use of this. i showed images above. it sounds like you really just dont want this. you can reject it, people can just use some other library or toolkit to make it happen

**Comment by bghira** [comment]
fwiw this even allows using the sdxl refiner to complete inference on sd 1.5 or 2.x

**Comment by patrickvonplaten** [comment]
> fwiw this even allows using the sdxl refiner to complete inference on sd 1.5 or 2.x

Can you add a quick code snippet for this?

**Comment by bghira** [comment]
i don't see why that's needed, it's the same thing that the SDXL pipeline uses, with the exception that you have to VAE decode the SD 2.x/1.x outputs:

```py
from diffusers import DiffusionPipeline
import torch

base = DiffusionPipeline.from_pretrained(
    "stabilityai/stable-diffusion-2-1", torch_dtype=torch.float16, variant="fp16", use_safetensors=True
).to("cuda")

refiner = DiffusionPipeline.from_pretrained(
    "stabilityai/stable-diffusion-xl-refiner-1.0",
    text_encoder_2=base.text_encoder_2,
    vae=base.vae,
    torch_dtype=torch.float16,
    use_safetensors=True,
    variant="fp16",
).to("cuda")

prompt = "A majestic lion jumping from a big stone at night"

base_image = base(
    prompt=prompt,
    num_inference_steps=40,
    denoising_end=0.8,
    output_type="pil",
).images[0]
image = refiner(
    prompt=prompt,
    num_inference_steps=40,
    denoising_start=0.8,
    image=base_image,
).images[0]
base_image.save('base_image.png', format='PNG')
image.save('image.png', format='PNG')
```

**Comment by bghira** [comment]
to use a SD 2.x zero-terminal SNR checkpoint that is finetuned for steps 0-400:

```py
from diffusers import DiffusionPipeline
import torch

base = DiffusionPipeline.from_pretrained(
    "stabilityai/stable-diffusion-2-1", torch_dtype=torch.float16, variant="fp16", use_safetensors=True
).to("cuda")

refiner = DiffusionPipeline.from_pretrained(
    "ptx0/pseudo-flex-base",
    torch_dtype=torch.float16,
    use_safetensors=True,
).to("cuda")

prompt = "A majestic lion jumping from a big stone at night"

base_image = base(
    prompt=prompt,
    num_inference_steps=40,
    denoising_end=0.8,
    output_type="latent",
).images
image = refiner(
    prompt=prompt,
    num_inference_steps=40,
    denoising_start=0.8,
    image=base_image,
).images[0]
base_image.save('base_image.png', format='PNG')
image.save('image.png', format='PNG')
```

File: src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl_img2img.py [hunk 1]
```diff
@@ -811,7 +810,7 @@ def __call__(
                 final 20% of the scheduler. The denoising_end parameter should ideally be utilized when this pipeline
                 forms a part of a "Mixture of Denoisers" multi-pipeline setup, as elaborated in [**Refine Image
                 Quality**](https://huggingface.co/docs/diffusers/using-diffusers/sdxl#refine-image-quality).
-            guidance_scale (`float`, *optional*, defaults to 7.5):
+            guidance_scale (`float`, *optional*, defaults to 5.0):
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl_img2img.py:813]: nice catch!

**Comment by patrickvonplaten** [comment]
I'm not getting great results with https://github.com/huggingface/diffusers/pull/4355#issuecomment-1906728628 and https://github.com/huggingface/diffusers/pull/4355#issuecomment-1906736194 doesn't work on this branch for me.

I'm ok adding it though, it makes sense to have this functionality when we already have some refiner model for SD2 1 trained.

**Comment by patrickvonplaten** [comment]
@yiyixuxu could you give this a review?

**Comment by bghira** [comment]
ah, yeah to be fair i haven't pulled this branch in some time, it's not very easy for me to test this stuff locally as i'm in central america and downloading these models takes a very long time, if it completes at all.

once it is in though, i can revisit it and put some compute toward training a specific refiner for this case.

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py [hunk 2]
```diff
@@ -524,16 +524,33 @@ def check_inputs(
                     f" {negative_prompt_embeds.shape}."
                 )
 
-    def get_timesteps(self, num_inference_steps, strength, device):
+    def get_timesteps(self, num_inference_steps, strength, device, denoising_start=None):
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py:527]: Can we add a `#copy from` statement here - I think here we can use same `get_timesteps` function in SDXL img2img, no?

note that the SDXL img2img `get_timesteps` implementation include code to make sure mixure-of-exports work correctly with 2nd order scheduler (see PR here https://github.com/huggingface/diffusers/pull/5511)
```suggestion
     # Copied from diffusers.pipelines.stable_diffusion_xl.pipeline_stable_diffusion_xl.StableDiffusionXLImg2ImgPipeline.get_timesteps
    def get_timesteps(self, num_inference_steps, strength, device, denoising_start=None):
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py [hunk 3]
```diff
@@ -747,21 +786,60 @@ def __call__(
         # 4. Preprocess image
         image = self.image_processor.preprocess(image)
 
-        # 5. set timesteps
+        # 5. Prepare timesteps
+        def denoising_value_valid(dnv):
+            return type(denoising_end) == float and 0 < dnv < 1
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py:791]: Is this what you want here instead?
```suggestion
            return type(dvd) == float and 0 < dnv < 1
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py [hunk 4]
```diff
@@ -747,21 +786,60 @@ def __call__(
...
+
         self.scheduler.set_timesteps(num_inference_steps, device=device)
         timesteps, num_inference_steps = self.get_timesteps(num_inference_steps, strength, device)
+        timesteps, num_inference_steps = self.get_timesteps(
+            num_inference_steps, strength, device, denoising_start=denoising_start if denoising_value_valid else None
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py:796]: `denoising_value_valid` is a function here, is this what we want? 
```suggestion
            num_inference_steps, strength, device, denoising_start=denoising_start if denoising_value_valid(denoising_start) else None
```

this PR is related too https://github.com/huggingface/diffusers/issues/6495

File: src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl_img2img.py [hunk 5]
```diff
@@ -787,12 +787,11 @@ def __call__(
...
-                will be used as a starting point, adding more noise to it the larger the `strength`. The number of
-                denoising steps depends on the amount of noise initially added. When `strength` is 1, added noise will
-                be maximum and the denoising process will run for the full number of iterations specified in
-                `num_inference_steps`. A value of 1, therefore, essentially ignores `image`. Note that in the case of
-                `denoising_start` being declared as an integer, the value of `strength` will be ignored.
+                Indicates extent to transform the reference `image`. Must be between 0 and 1. `image` is used as a
+                starting point and more noise is added the higher the `strength`. The number of denoising steps depends
+                on the amount of noise initially added. When `strength` is 1, added noise is maximum and the denoising
+                process runs for the full number of iterations specified in `num_inference_steps`. A value of 1
+                essentially ignores `image`.
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl_img2img.py:794]: why do we edit this? and removed the mention of this
 > Note that in the case of `denoising_start` being declared as an integer, the value of `strength` will be ignored.

**Review by yiyixuxu — COMMENTED** [review]
thanks!
left a few questions, I think what's left is just a test :)

File: src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl_img2img.py (hunk 5 above)
Comment (bghira) [reply src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl_img2img.py:794]: keep in mind that this PR predates the other changes

**Comment by sayakpaul** [comment]
Not stale.

**Comment by yiyixuxu** [comment]
@bghira 
do we still want to support this? happy to help with the PR but it only makes sense if we are going to have the checkpoints though

**Comment by bghira** [comment]
yes but any sd model can be an expert. pick one for composition and one for details or style and split the job between them

**Comment by yiyixuxu** [comment]
@bghira 

would you be able to provide examples so we can make a doc page for this?

Code-wise, I think there's not much left to finish up. Do you want to finish it up, or would you prefer that I take this over this PR?

**Comment by bghira** [comment]
@yiyixuxu actually i'll be able to put more effort forth on this soon, thanks for the patience

**Comment by Shinyzenith** [comment]
Hi, is there anything blocking this pr? I'd love to try and help with it.

**Comment by bghira** [comment]
it needs to be re-tested with the ip adapter and other additions that have come around since the initial SDXL implementation of this function.

i've updated the PR to resolve conflicts but i haven't looked much deeper at whether these are the correct solutions yet

**Comment by bghira** [comment]
coming up on a happy 1 year anniversary to those following this PR :)

**Comment by bghira** [comment]
closing due to lack of interest, use ComfyUI for this workflow i guess
